src/java.desktop/share/classes/java/awt/SystemTray.java

```diff
@@ -343,7 +345,7 @@ public TrayIcon[] getTrayIcons() {
         @SuppressWarnings("unchecked")
         Vector<TrayIcon> icons = (Vector<TrayIcon>)AppContext.getAppContext().get(TrayIcon.class);
         if (icons != null) {
-            return icons.toArray(new TrayIcon[icons.size()]);
+            return icons.toArray(new TrayIcon[0]);
         }
```

Comment:
It is not intuitive but the "zero version" is actually faster:

Reply:
ok ... why ?

Reply:
It's explained in Aleksey Shipilёv's article
https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion

Reply:
In fact, there's no need to create zero-sized array, there's one already: `EMPTY_TRAY_ARRAY`.

Yet the code may look weird with it.

```java
return icons.toArray(EMPTY_TRAY_ARRAY);
```

Reply:
We probably can extract this to the separate cleanup, the resulted code is smaller and faster.
PS I have added a link to my previous message accidentally hit submit button.

Reply:
I'm sure I've used this pattern
return icons.toArray(EMPTY_TRAY_ARRAY);
in code that runs often and may be performance sensitive, so I don't see anything weird about it.
OTOH I am not sure this code is anything that needs super-optimising .. we seem to have gone down that discussion route because it was the justification for the change.

BTW the shiplev article, which I skimmed must be suggesting that the array initialisation must be faster if it is on the array it allocates itself because of the APIs used .. but as he also notes things can change.

Anyway this is mildly interesting discussion but no one will ever, ever be able to measure a difference when it used here since it is probably 0.0001 % of the work ...

Reply:
It also has a functional implication, in the old code  the "size" on a Vector is called outside of the lock. So the resulted array after "icons.toArray" can be bigger than the number of icons. This is similar to this:
https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.desktop/share/classes/javax/sound/midi/Sequence.java#L281
I think that race can even be triggered by the new test.

Reply:
I remember IDEs used to suggest using correctly sized arrays in the calls to `toArray`. Now they suggest using zero-sized array.

My idea was to resolve the warning raised by the IDE to reduce the number of yellow highlights on the file outline.

I agree the performance gain if any is negligible in this case: we don't expect many icons, probably no more than 3.

Reply:
The IDE warning was changed based on the article above. Did you have a chance to reproduce a possible race caused by data change after the size() was called?

Reply:
I can reproduce it by this code example:
```
class Scratch {

    static volatile Throwable failed;
    static volatile long endtime;

    public static void main(String[] args) throws Exception {
        if (!SystemTray.isSupported()) {
            return; // passed if SystemTray is not supported
        }
        SystemTray st = SystemTray.getSystemTray();

        endtime = System.nanoTime() + TimeUnit.SECONDS.toNanos(5);

        Thread thread1 = new Thread(() -> {
            while (!isComplete()) {
                try {
                    st.add(new TrayIcon(new BufferedImage(10, 10,
                                                          BufferedImage.TYPE_INT_ARGB)));
                } catch (AWTException e) {
                    failed = e;
                }
            }
        });
        Thread thread2 = new Thread(() -> {
            while (!isComplete()) {
                try {
                    st.add(new TrayIcon(new BufferedImage(10, 10,
                                                          BufferedImage.TYPE_INT_ARGB)));
                } catch (AWTException e) {
                    failed = e;
                }
            }
        });
        Thread thread3 = new Thread(() -> {
            while (!isComplete()) {
                TrayIcon[] icons = st.getTrayIcons();
                if (icons.length > 0) {
                    st.remove(icons[0]);
                }
            }
        });
        Thread thread4 = new Thread(() -> {
            while (!isComplete()) {
                TrayIcon[] icons = st.getTrayIcons();
                if (icons.length > 0 && icons[icons.length - 1] == null) {
                    failed = new RuntimeException(
                            "icon is null:" + Arrays.toString(icons));
                }
            }
        });
        thread1.start();
        thread2.start();
        thread3.start();
        thread4.start();
        thread1.join();
        thread2.join();
        thread3.join();
        thread4.join();
        for (TrayIcon icon : st.getTrayIcons()) {
            st.remove(icon); // cleanup
        }
        if (failed != null) {
            System.err.println("Test failed");
            throw new RuntimeException(failed);
        }
    }

    private static boolean isComplete() {
        return endtime - System.nanoTime() < 0 || failed != null;
    }
}
```

Reply:
BTW I think that the usage of 
`return icons.toArray(EMPTY_TRAY_ARRAY);`
is slower than any of other cases: presized or zero. So if we would like to use a similar pattern the zero version is better.

Reply:
> BTW I think that the usage of
> `return icons.toArray(EMPTY_TRAY_ARRAY);`
> is slower than any of other cases: presized or zero. So if we would like to use a similar pattern the zero version is better.

Why is it? It's the same as `icons.toArray(new TrayIcon[0])` without creating a new zero-sized array.

Reply:
It should be checked of course, but in zero case the size is known by the compiler not sure that in the cached case it can make the same optimizations. The article above said the cached case is a "marginal improvement over new Foo[0]"

Reply:
I'm unsure if the Java compiler performs any optimisations here. The called `toArray` verifies the size first, and it's the same size either way, which is zero.

If there were no pre-allocated array, I wouldn't bother with a cached copy. But we have one. One less object to be allocated and to be garbage-collected.

Since the number of the icons is unlikely to become large, I doubt there'll be any noticeable difference.

As your sample above demonstrates, the old code could return an array with a null element in it. This is resolved by the new code.